### ISSUE TITLE
Implicit and higher-order time-stepping schemes

### ISSUE THREAD

**Issue by glwagner** [glwagner/PriceWellerPinkel.jl#5]
We need new time-stepping schemes: both implicit schemes and higher-order schemes (currently we only have ForwardEuler). 

To implement any implicit scheme we will need (at least) a tridiagonal solver.

Perhaps the right place to start is Crank-Nicholson for simple diffusion.

Further down the line, we may consider [Implicit-Explicit Runge Kutta schemes](https://ac.els-cdn.com/S0168927497000561/1-s2.0-S0168927497000561-main.pdf?_tid=9ba20584-7940-49a7-8d6a-de6a1e919d21&acdnat=1552010568_6528ebffbbccd5e3d54b50a8faf13981).

A main question is how to abstract the definition of a linear equation. A decent starting place is to assume that only a term of the 'diffusive' type will need to be treated implicitly.

cc @SandreOuza

**Comment by sandreza** [comment]
Some comments:

1) Julia has built in Tridiagonal matrices with efficient solves!

2) I agree that the simplest starting point is to let each variable have its own implicit tridiagonal matrix as opposed to creating a general linear matrix inversion that involves all the other solution variables.

- Andre

**Comment by glwagner** [comment]
@SandreOuza do you have a reference for the solvers? I can't find info. 

I see here:

https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/index.html#LinearAlgebra.SymTridiagonal

They refer to an 'efficient eigensolver'. But I don't understand... I thought we can solve in O(n) with Gaussian elimination. Why do we need an eigensolver? It's also only for symmetric tridiagonal matrices apparently. Why, again, is that?

**Comment by sandreza** [comment]
@glwagner 

I just meant that it \ operation (or better yet ldiv! ) uses the LAPACK implementation for tridiagonal matrices (aka Gaussian elimination). In my tests it is about 30% or faster than my own naive implementation of Thomas' algorithm.

**Comment by sandreza** [comment]
@glwagner 

more specifically I mean something like
 up = zeros(n-1)
 lo = zeros(n-1)
 di = ones(n)
 y = lu(Tridiagonal(up,di,lo))

computes the lu factorization of a tridiagonal matrix with super diagonals up, diagonals di, and sub diagonals lo. From there one can use ldiv!(y,x) to compute the solution to the matrix system yc = x

**Comment by glwagner** [comment]
There’s nothing like a test! I get what you’re saying, I feel a bit mystified due to lack of information. Maybe I need to read the LAPACK docs...

Sounds like ldiv is our starting point either way. As it should be.

**Comment by glwagner** [comment]
@SandreOuza here's a script:

```julia
using LinearAlgebra, Random, BenchmarkTools

function get_Axb(n)

    d = rand(n)
    dl = rand(n-1)
    du = rand(n-1)
    A = Tridiagonal(dl, d, du) 
    x = rand(n)
    b = rand(n)

    return A, x, b
end

function solve_tridiagonal!(A, x, b, nloops)
    for i = 1:nloops
        x .= A \ b 
    end 

    return nothing
end

for n in 2 .^ [5, 6, 7]

    A, x, b = get_Axb(n)

    solve_tridiagonal!(A, x, b, 1)

    @show n

    @btime solve_tridiagonal!($A, $x, $b, 1)
    @btime solve_tridiagonal!($A, $x, $b, 10) 
    @btime solve_tridiagonal!($A, $x, $b, 100)

end
```

The results:

```julia
julia> include("test.jl")
n = 32
  768.088 ns (9 allocations: 2.41 KiB)
  7.545 μs (90 allocations: 24.06 KiB)
  76.007 μs (900 allocations: 240.63 KiB)
n = 64
  1.309 μs (9 allocations: 3.73 KiB)
  12.975 μs (90 allocations: 37.34 KiB)
  131.232 μs (900 allocations: 373.44 KiB)
n = 128
  2.457 μs (9 allocations: 6.89 KiB)
  24.640 μs (90 allocations: 68.91 KiB)
  248.936 μs (900 allocations: 689.06 KiB)
```

Long story short, I think you're right --- and it looks like we don't need `ldiv!`; we can just use dot-syntax with `\`. Also, it seems that the solver for `Tridiagonal` makes exactly 9 allocations per solve (we can live with that, but its interesting!)

**Comment by glwagner** [comment]
My idea: let's define a new time-stepper called, for starters, `BackwardEuler`:

```julia
struct BackwardEulerTimeStepper{T, G, K, TA} <: Timestepper
    rhs :: T
    A   :: TA
end
```
`rhs` is `similar(solution)` and `A` is a `FieldVector` of `Tridiagonal` matrices for each component of the solution `i`:

```julia
A[i] = I - Δt D κ[i] D
```
where `I` is the identity matrix, `Δt` is time-step, and `D` is the second-order matrix vertical derivative operator that assumes a no flux boundary condition.

Backward Euler then solves

```
A cⁿ⁺¹ = cⁿ
```

where `cⁿ` is the solution at iteration `n`.

The struct `BackwardEuler` thus has all the information needed to  construct `A` and solve for `cⁿ⁺¹`.
    
For inhomogeneous boundary conditions, we add information to the right-hand side of the matrix equation being solved every time-step.

**Comment by glwagner** [comment]
Next I think it would be nice to implement "SBDF" defined in equation 17 in [Ascher et al 1995](https://epubs.siam.org/doi/pdf/10.1137/0732037?casa_token=DvS-9Fxvu6oAAAAA:syFuNZZmwxg_tQ3utCqg-kYSRJvqLLVvvcD1zdXUguQ45_vSFvTS2yK8xQrz3c_KYlM1kz1g2g).

**Comment by glwagner** [comment]
A bit more: the equation must specify how to calculate the diffusivity `Κ` for each component of the solution. 

Here is some sample code that computes `A`:

```julia
# Flux divergence operator on `face`, computed over over `cell`
flux_div_op(m, κ, face, cell) = κ(m, face) / (Δc(m, face) * Δf(m, cell))

# Build backward Euler operator for diffusive problems
for i = 2:N-1
    # Lower diagonal
    A.dl[i-1] = -Δt*flux_div_op(m, K, i, i)

    # Upper diagonal
    A.du[i]   = -Δt*flux_div_op(m, K, i+1, i)

    # Center diagonal
    A.d[i]   = 1 + Δt*(flux_div_op(m, K, i+1, i) + flux_div_op(m, K, i, i))
end

A.du[1] = -Δt*flux_div_op(m, K, 2, 1)
A.d[1] = 1 + Δt*flux_div_op(m, K, 2, 1)

A.dl[end] = -Δt*flux_div_op(m, K, N, N)
A.d[end] = 1 + Δt*flux_div_op(m, K, N, N)
```

The backward Euler update for `ϕ` could then be

```julia
# compute rhs...
@. rhs.data += ϕ.data
ϕ.data .= rhs.data \ A
```

**Comment by sandreza** [comment]
That seems like a good start to me. If we ever want to be more flexible on how we do our implicit timestepping schemes (by coupling different fields together in the linear terms) then we may need to slightly modify things, but overall this is a good plan.

**Comment by glwagner** [comment]
Closed by #11.
